tools/travis_controller.sh

```diff
@@ -0,0 +1,211 @@
+        EXIT_CODE=$?
+    else
+        printf "\n==============================================================================\n"
+        printf "Previous build failure detected, skipping dependency-convergence check.\n"
+        printf "==============================================================================\n"
+    fi
+    
+    if [ $EXIT_CODE == 0 ]; then
+        check_shaded_artifacts
+        EXIT_CODE=$(($EXIT_CODE+$?))
+        check_shaded_artifacts_s3_fs hadoop
+        EXIT_CODE=$(($EXIT_CODE+$?))
+        check_shaded_artifacts_s3_fs presto
+        EXIT_CODE=$(($EXIT_CODE+$?))
+        check_shaded_artifacts_connector_elasticsearch ""
+        EXIT_CODE=$(($EXIT_CODE+$?))
+        check_shaded_artifacts_connector_elasticsearch 2
+        EXIT_CODE=$(($EXIT_CODE+$?))
+        check_shaded_artifacts_connector_elasticsearch 5
+        EXIT_CODE=$(($EXIT_CODE+$?))
+    else
+        echo "=============================================================================="
+        echo "Previous build failure detected, skipping shaded dependency check."
+        echo "=============================================================================="
+    fi
+
+    if [ $EXIT_CODE == 0 ]; then
+        echo "Creating cache build directory $CACHE_FLINK_DIR"
+        mkdir -p "$CACHE_FLINK_DIR"
+    
+        cp -r . "$CACHE_FLINK_DIR"
+
+        function minimizeCachedFiles() {
+            # reduces the size of the cached directory to speed up
+            # the packing&upload / download&unpacking process
+            # by removing files not required for subsequent stages
+    
+            # original jars
+            find "$CACHE_FLINK_DIR" -maxdepth 8 -type f -name 'original-*.jar' | xargs rm -rf
+    
+            # .git directory
+            # not deleting this can cause build stability issues
+            # merging the cached version sometimes fails
+            rm -rf "$CACHE_FLINK_DIR/.git"
+        }
+    
+        start_fold "minimize_cache" "Minimizing cache"
+        travis_time_start
+        minimizeCachedFiles
+        travis_time_finish
+        end_fold "minimize_cache"
+    else
+        echo "=============================================================================="
+        echo "Previous build failure detected, skipping cache setup."
+        echo "=============================================================================="
+    fi
+elif [ $STAGE != "$STAGE_CLEANUP" ]; then
+	if ! [ -e $CACHE_FLINK_DIR ]; then
+		echo "Cached flink dir $CACHE_FLINK_DIR does not exist. Exiting build."
+		exit 1
+	fi
+	# merged compiled flink into local clone
+	# this prevents the cache from being re-uploaded
+	start_fold "merge_cache" "Merging cache"
+	travis_time_start
+	cp -RT "$CACHE_FLINK_DIR" "."
+	travis_time_finish
+	end_fold "merge_cache"
+
+	start_fold "adjust_timestamps" "Adjusting timestamps"
+	travis_time_start
+	# adjust timestamps to prevent recompilation
+	find . -type f -name '*.java' | xargs touch
```

Comment:
Why is this necessary to prevent recompilation? Does the compiler plugin recompile everything if the sources are old enough?

Reply:
the compiler compares the modification time of the source files and compiled files. If the mod time of the source time is newer it recompiles the class. Here we ensure that the mod time of the source files is older than the compiled files.

Reply:
the `.timestamp` is used by the scala compiler plugin btw

Reply:
By `touch`ing all files, we set the mod time of all source files to _"now"_. How is this older than the compiled files?